server/mergin/sync/models.py

```diff
@@ -902,6 +905,250 @@ def construct_checkpoint(self) -> bool:
+                    rank=checkpoint.rank,
+                    version=checkpoint.end,
+                ).first()
+
+                if not existing_diff_checkpoint:
+                    base_file = FileHistory.get_basefile(file_path_id, checkpoint.end)
+                    if not base_file:
+                        continue
+
+                    diff_path = mergin_secure_filename(
+                        f"{item.path}-diff-{uuid.uuid4()}"
+                    )
+                    checkpoint_diff = FileDiff(
+                        basefile=base_file,
+                        path=diff_path,
+                        rank=checkpoint.rank,
+                        version=checkpoint.end,
+                    )
+                    # Patch the delta with the path to the new diff checkpoint
+                    item.diff = diff_path
+                    db.session.add(checkpoint_diff)
+
+        checkpoint_change = ProjectVersionChange(
+            version_id=changes_range[-1].version_id,
+            rank=checkpoint.rank,
+            delta=DeltaDataSchema(many=True).dump(merged_delta_items),
+        )
+        db.session.add(checkpoint_change)
+        db.session.commit()
+        return checkpoint_change
+
+    @classmethod
+    def query_changes(cls, project_id, since, to, rank=None):
+        """Query changes with specified parameters"""
+        query = cls.query.join(ProjectVersion).filter(
+            ProjectVersion.project_id == project_id,
+            ProjectVersion.name >= since,
+            ProjectVersion.name <= to,
+        )
+        if rank is not None:
+            query = query.filter(ProjectVersionChange.rank == rank)
+        return query.order_by(ProjectVersion.name).all()
+
+    @classmethod
+    def get_delta(
+        cls, project_id: str, since: int, to: int
+    ) -> Optional[List[DeltaMerged]]:
+        """
+        Get changes between two versions, merging them if needed.
+        - create FileDiff checkpoints if needed
+        - create ProjectVersionChange checkpoints if needed with delta json
+        """
+        if since > to:
+            logging.error(
+                f"Start version {since} is higher than end version {to} - broken history"
+            )
+            return
+        if since == to:
+            # Return only changes for this version
+            change = (
+                ProjectVersionChange.query.join(ProjectVersion)
+                .filter(
+                    ProjectVersion.project_id == project_id,
+                    ProjectVersion.name == to,
+                    ProjectVersionChange.rank == 0,
+                )
+                .first()
+            )
+            return [DeltaMerged(**item) for item in change.delta] if change else None
+
+        expected_checkpoints = Checkpoint.get_checkpoints(since + 1, to)
+        expected_changes: List[ProjectVersionChange] = (
+            ProjectVersionChange.query.join(ProjectVersion)
```

Comment:
hm, given the usage (and general clarity) it might be better to move from `version_id` to `project_id` + `version_name` instead...